exec/java-exec/src/test/java/org/apache/drill/TestTpchDistributedConcurrent.java

```diff
@@ -0,0 +1,206 @@
+    final String filename = queryFile[random.nextInt(queryFile.length)];
+    final String query;
+    try {
+      query = QueryTestUtil.normalizeQuery(getFile(filename)).replace(';', ' ');
+    } catch(IOException e) {
+      throw new RuntimeException("Caught exception", e);
+    }
+    final UserResultsListener listener = new ChainingSilentListener(query);
+    client.runQuery(UserBitShared.QueryType.SQL, query, listener);
+    synchronized(this) {
+      listeners.put(listener, listener);
+    }
+  }
+
+  private class ChainingSilentListener extends SilentListener {
+    private final String query;
+
+    public ChainingSilentListener(final String query) {
+      this.query = query;
+    }
+
+    @Override
+    public void queryCompleted(QueryState state) {
+      super.queryCompleted(state);
+
+      completionSemaphore.release();
+      synchronized(TestTpchDistributedConcurrent.this) {
+        final Object object = listeners.remove(this);
+        assertNotNull("listener not found", object);
+
+        /* Only submit more queries if there hasn't been an error. */
+        if ((failedQueries.size() == 0) && (remainingQueries > 0)) {
+          /*
+           * We can't directly submit the query from here, because we're on the RPC
+           * thread, and it throws an exception if we try to send from here. So we
+           * allow the QuerySubmitter thread to advance.
+           */
+          submissionSemaphore.release();
+          --remainingQueries;
+        }
+      }
+    }
+
+    @Override
+    public void submissionFailed(UserException uex) {
+      super.submissionFailed(uex);
+
+      completionSemaphore.release();
+      System.out.println("submissionFailed for " + query + "\nwith " + uex);
+      synchronized(TestTpchDistributedConcurrent.this) {
+        final Object object = listeners.remove(this);
+        assertNotNull("listener not found", object);
+        failedQueries.add(new FailedQuery(query, uex));
+        testThread.interrupt();
+      }
+    }
+  }
+
+  private class QuerySubmitter extends Thread {
+    @Override
+    public void run() {
+      while(true) {
+        try {
+          submissionSemaphore.acquire();
+        } catch(InterruptedException e) {
+          System.out.println("QuerySubmitter quitting.");
+          return;
+        }
+
+        submitRandomQuery();
+      }
+    }
+  }
+
+  @Test
+  public void testConcurrentQueries() throws Exception {
+    QueryTestUtil.testRunAndPrint(client, UserBitShared.QueryType.SQL, alterSession);
+
+    testThread = Thread.currentThread();
+    final QuerySubmitter querySubmitter = new QuerySubmitter();
+    querySubmitter.start();
+
+    // Kick off the initial queries. As they complete, they will submit more.
+    submissionSemaphore.release(CONCURRENT_QUERIES);
+
+    // Wait for all the queries to complete.
+    try {
+      completionSemaphore.acquire(TOTAL_QUERIES);
+    } catch(InterruptedException e) {
+      fail("caught " + e);
```

Comment:
If ```testThread``` is interrupted, this fail will trigger an ```AssertionError("...")```, however ```querySubmitter``` is not interrupted, and the ```System.err.println("...") ``` calls do not happen, right?